source/credentials_provider_imds.c

```diff
@@ -333,9 +386,45 @@ static void s_imds_on_stream_complete_fn(struct aws_http_stream *stream, int err
     struct aws_credentials_provider_imds_impl *impl = imds_user_data->imds_provider->impl;
     impl->function_table->aws_http_stream_release(stream);
 
+    if (imds_user_data->query_state == AWS_IMDS_QS_TOKEN) {
+        /* Gets 400 means token is required but the request itself failed. */
+        if (imds_user_data->status_code == 400) {
+            s_imds_finalize_get_credentials_query(imds_user_data);
+            return;
+        }
+
+        /*
+         * Other than that, if meets any error, then token is not required,
+         * we should fall back to insecure request. Otherwise, we should use
+         * token in following requests.
+         */
+        if (imds_user_data->status_code != 200 || imds_user_data->current_result.len == 0) {
+            impl->token_required = false;
+        } else {
+            imds_user_data->token = aws_byte_cursor_from_buf(&(imds_user_data->current_result));
```

Comment:
Doesn't current_result get cleared?  Token needs to be used for up to two more queries right?